### PR TITLE
Implementation of data structure verification

### PR DESCRIPTION
**Pull request by hcaw**
Included is the complete implementation of both the one-dimensional
list and Merkle tree data structures (with both single leaf and all-leaf
verification for the Merkle tree structure).

The result of gas testing shows the following figures, for a list of 8
items:
i. verifyMessageArray (all items verified): `28505`
ii. verifyMerkleAll: `29728`
iii. verifyMerkleLeaf: `25423`

### CODE REVIEW

File: contracts/Verification.sol [hunk 1]
```diff
@@ -1,100 +1,88 @@
...
+
+        for (uint256 j = 0; hashLength > 1; j = 0) {
+            for (uint256 i = 0; i < hashLength; i = i + 2) {
+                _data[j] = keccak256(abi.encodePacked(_data[i], _data[i + 1]));
+                j = j + 1;
+            }
+            // This effectively halves the list length every time,
+            // but a subtraction op-code costs less
+            hashLength = hashLength - j;
+        }
```
Comment (philipstanislaus) [inline contracts/Verification.sol:54]: Makes sense – could you just add a comment on top about the purpose of each loop? That might make this a bit easier to grasp. Thanks 👍

File: contracts/Verification.sol [hunk 2]
```diff
@@ -1,100 +1,88 @@
...
+        bytes32 root,
+        bytes32 leaf,
+        bytes32[] calldata proof
+    ) public pure returns (bool) {
+        bytes32 computedHash = leaf;
+
+        for (uint256 i = 0; i < proof.length; i++) {
+            bytes32 proofElement = proof[i];
+
+            if (computedHash < proofElement) {
```
Comment (philipstanislaus) [inline contracts/Verification.sol:79]: What's the gas cost of this check compared to adding a bitfield to tell this function whether an element should be left or right in each hash?

**Review by philipstanislaus — COMMENTED** [review]
Nice, thanks! 

Could you run the gas estimates again and add numbers into a table for the following message counts please? 

1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192

That would be amazing. Many thanks 🎉

File: contracts/Verification.sol (hunk 2 above)
Comment (hcaw) [reply contracts/Verification.sol:79]: I'd say slightly less, given that the current implementation would be:
(If `n` is _number of elements_)
`n * (less_than_opcode + equality_operator) = 6n`
If you used a bitfield to encode the left/right property for each element, you would only require a single conditional operator:
`n * equality_operator = 3n`
However the bitfield itself would require an additional function parameter of type `bytes` which would needed to be loaded onto the stack from calldata, which costs roughly:
`(n * calldataload)/32`

**Comment by hcaw** [comment]
> Nice, thanks!
> 
> Could you run the gas estimates again and add numbers into a table for the following message counts please?
> 
> 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192
> 
> That would be amazing. Many thanks 🎉

Pasting here from Slack:

```
verifyMessageArray function:
  ↪ Data length 1:  22991
  ↪ Data length 2:  23782
  ↪ Data length 4:  25364
  ↪ Data length 8:  28517
  ↪ Data length 16:  34848
  ↪ Data length 32:  47506
  ↪ Data length 64:  72789
  ↪ Data length 128:  123523
  ↪ Data length 256:  225470
  ↪ Data length 512:  431020
  ↪ Data length 1024:  848984
  ↪ Data length 2048:  1712848
  ↪ Data length 4096:  3551048
  ↪ Data length 8192:  7669348
verifyMerkleLeaf function:
  ↪ Data length 1:  22930
  ↪ Data length 2:  23783
  ↪ Data length 4:  24602
  ↪ Data length 8:  25435
  ↪ Data length 16:  26262
  ↪ Data length 32:  27125
  ↪ Data length 64:  27959
  ↪ Data length 128:  28798
  ↪ Data length 256:  29641
  ↪ Data length 512:  30429
  ↪ Data length 1024:  31344
  ↪ Data length 2048:  32097
  ↪ Data length 4096:  32987
  ↪ Data length 8192:  33816
verifyMerkleAll function:
  ↪ Data length 1:  23100
  ↪ Data length 2:  24102
  ↪ Data length 4:  26015
  ↪ Data length 8:  29728
  ↪ Data length 16:  37125
  ↪ Data length 32:  51769
  ↪ Data length 64:  81098
  ↪ Data length 128:  139760
  ↪ Data length 256:  257869
  ↪ Data length 512:  496948
  ↪ Data length 1024:  987039
  ↪ Data length 2048:  2016990
  ↪ Data length 4096:  4272917
  ↪ Data length 8192: _(reverted due to costing > 9.5m gas)_
```

File: contracts/Verification.sol (hunk 2 above)
Comment (hcaw) [reply contracts/Verification.sol:79]: Just had a thought, the last calculation is actually `(n * calldataload)/256`, which makes the bitfield cost a bit less than I originally thought

**Pull request merged**